openverse_catalog/dags/providers/provider_api_scripts/inaturalist.py

```diff
@@ -0,0 +1,177 @@
+            except for adding ancestry tags to the taxa table.
+"""
+import os
+from pathlib import Path
+from textwrap import dedent
+from typing import Dict
+
+import pendulum
+from airflow.exceptions import AirflowSkipException
+from airflow.operators.python import PythonOperator
+from airflow.providers.amazon.aws.hooks.s3 import S3Hook
+from airflow.providers.postgres.hooks.postgres import PostgresHook
+from airflow.providers.postgres.operators.postgres import PostgresOperator
+from airflow.utils.task_group import TaskGroup
+from common.constants import POSTGRES_CONN_ID
+from common.licenses import LicenseInfo, get_license_info
+from common.loader import provider_details as prov
+from providers.provider_api_scripts.provider_data_ingester import ProviderDataIngester
+
+
+AWS_CONN_ID = os.getenv("AWS_CONN_ID", "test_conn_id")
+PROVIDER = prov.INATURALIST_DEFAULT_PROVIDER
+SCRIPT_DIR = Path(__file__).parents[1] / "provider_csv_load_scripts/inaturalist"
+SOURCE_FILE_NAMES = ["photos", "observations", "taxa", "observers"]
+
+
+class iNaturalistDataIngester(ProviderDataIngester):
+
+    providers = {"image": prov.INATURALIST_DEFAULT_PROVIDER}
+
+    def __init__(self, *kwargs):
+        super(iNaturalistDataIngester, self).__init__()
+        self.pg = PostgresHook(POSTGRES_CONN_ID)
+
+        # adjustments to buffer limits. TO DO: try to integrate this with the dev
+        # environment logic in the base class, rather than just over-writing it.
+        self.media_stores["image"].buffer_length = 10_000
+        self.batch_limit = 10_000
+
+        self.sql_template = dedent(
+            (SCRIPT_DIR / "05_export_to_json_template.sql")
+            .read_text()
+            .replace("batch_limit", str(self.batch_limit))
+        )
+
+    def get_next_query_params(self, prev_query_params=None, **kwargs):
+        if prev_query_params is None:
+            return {"offset_num": 0}
+        else:
+            next_offset = prev_query_params["offset_num"] + self.batch_limit
+            return {"offset_num": next_offset}
+
+    def get_response_json(self, query_params: Dict):
+        """
+        Call the SQL to pull json from Postgres, where the raw data has been loaded.
+        """
+        offset_num = str(query_params["offset_num"])
+        sql_string = self.sql_template.replace("offset_num", offset_num)
+        return self.pg.get_records(sql_string)
+
+    def get_batch_data(self, response_json):
+        if response_json:
+            return [dict(item[0]) for item in response_json]
+        return None
+
+    def get_record_data(self, data):
+        license_url = data.get("license_url")
+        license_info = get_license_info(license_url=license_url)
+        if license_info == LicenseInfo(None, None, None, None):
+            return None
+
+        foreign_id = data.get("foreign_id")
+        if foreign_id is None:
+            return None
+
+        return {
+            "foreign_identifier": f"{foreign_id}",
```

Comment:
Does this need to be interpolated?

```suggestion
            "foreign_identifier": foreign_id,
```

Fun factoid: I read recently that interpolating to an `f` string can be slightly faster than calling `str` on an object :scream:

https://web.archive.org/web/20190617081920/https://stackoverflow.com/questions/56587807/why-are-f-strings-faster-than-str-to-parse-values

Reply:
Wow! I think I was assuming that because the db field is `varchar` I should convert it, but I just tried it out and it works fine without the conversion. Thanks!